WordPress/openverse-frontend
# Update banner styles

## Fixes


Fixes #1123 by @panchovm

## Description


Primarily focused on updating the banner styles and unifying the implementations so that both have a close button. The unification effort is where most of the changes come from as well as some on-the-fly TypeScript conversion.

I'd like to split this PR into multiple parts but will be on vacation for the next week so just wanted to get my WIP up in case someone else wanted to pick it apart while I'm gone.

Changes proposed:
* Convert `use-storage` and `use-event-listener` to TypeScript and simplify the initial value handling in use-storage.
* Update `local.ts` to implement the full `Storage` API 1:1 instead of using a bespoke API. Also simplify the error handling in it.
* Convert NotificationBanner to TypeScript and update it to be able to handle dismissal on its own rather than relying on the parent to handle it. This makes it easy to implement new dismissable banners without having to duplicate logic.
* Update TranslationBanner to use the new NotificationBanner API and remove the unnecessary stuff from `use-i18n-sync` (which can probably be renamed now that nothing is being "synced")

## Testing Instructions

Checkout the branch and run `pnpm dev`. Visit the site in English and verify there is no banner. Visit the site in Russian  (`ru`) and verify there is a banner and that it can be dismissed and that it stays dismissed between requests (aside from the already existing problem that the banner is rendered SSR, to be fixed by the UI state cookie milestone).

Visit the site in another low-translation percent language like Portuguese (`pt`). Verify that a new banner shows with the correct information and that it is dismissable and stays dismissed. Verify that you can switch between Russian and Portuguese and get no banner.

Now do the same with two more low-translation-percent languages and do not dismiss the banners and ensure that both show the appropriate banner between page requests.

Note: All of the above could be encoded in e2e tests and I plan to do that when I get back from vacation and have time to implement additional tests for this on top of the already existing ones.

The CC referral banner is not yet implemented but testing it will go as follows:

Visit the site with the `referrer=creativecommons.org` query param and ensure that the banner shows and that it is dismissable. Ensure that it stays dismissed on subsequent requests to the site with the query param. (note again that this has not yet been implemented on this branch!).

## Checklist


[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
